core/modules/replica/DatabaseServicesMySQL.cc

```diff
@@ -521,114 +570,174 @@ void DatabaseServicesMySQL::saveReplicaInfoImpl(util::Lock const& lock,
+    findWorkerReplicasImpl(lock, oldReplicaInfoCollection, worker, database);
 
-        // Check each old replicas to see if it's present in the new collection
-        std::vector<ReplicaInfo> oldReplicas;
-        if (findWorkerReplicasImpl(lock, oldReplicas, worker, database)) {
+    LOGS(_log, LOG_LVL_DEBUG, context << "old replicas fetch: 2"
+         << " num.replicas: " << oldReplicaInfoCollection.size());
 
-            for (ReplicaInfo const& replica: oldReplicas) {
-                unsigned int const chunk = replica.chunk();
+    LOGS(_log, LOG_LVL_DEBUG, context << "old replicas group: 1");
 
-                // Eliminate the 'dead' chunks entry from the database
-                if (not newReplicas.at(worker).at(database).count(chunk)) {
-                    _conn->execute(
-                        "DELETE FROM " + _conn->sqlId(   "replica") +
-                        "  WHERE " +     _conn->sqlEqual("worker",   worker) +
-                        "    AND " +     _conn->sqlEqual("database", database) +
-                        "    AND " +     _conn->sqlEqual("chunk",    chunk));
-                }
+    WorkerDatabaseChunkMap<ReplicaInfo const*> oldReplicas;
+    for (auto&& replica: oldReplicaInfoCollection) {
+        oldReplicas.atWorker(replica.worker())
+                   .atDatabase(replica.database())
+                   .atChunk(replica.chunk()) = &replica;
+    }
+
+    LOGS(_log, LOG_LVL_DEBUG, context << "old replicas group: 2");
+
+    // Find differences between the collections
+
+    LOGS(_log, LOG_LVL_DEBUG, context << "*** replicas compa: 1");
+
+    WorkerDatabaseChunkMap<ReplicaInfo const*> inBoth;
+    SemanticMaps::intersect(newReplicas,
+                            oldReplicas,
+                            inBoth);
+
+    WorkerDatabaseChunkMap<ReplicaInfo const*> inNewReplicasOnly;
+    WorkerDatabaseChunkMap<ReplicaInfo const*> inOldReplicasOnly;
+    SemanticMaps::diff2(newReplicas,
+                        oldReplicas,
+                        inNewReplicasOnly,
+                        inOldReplicasOnly);
+
+    LOGS(_log, LOG_LVL_DEBUG, context << "*** replicas compa: 2"
+         << " #both: " << SemanticMaps::count(inBoth)
+         << " #new: " << SemanticMaps::count(inNewReplicasOnly)
+         << " #old: " << SemanticMaps::count(inOldReplicasOnly));
+
+    // Eiminate outdated replicas
+
+    LOGS(_log, LOG_LVL_DEBUG, context << "old replicas remov: 1");
+    
+    for (auto&& worker: inOldReplicasOnly.workerNames()) {
+
+        auto const& databases = inOldReplicasOnly.worker(worker);
+        for (auto&& database: databases.databaseNames()) {
+
+            auto const& chunks = databases.database(database);
+            for (auto&& chunk: chunks.chunkNumbers()) {
+                deleteReplicaInfoImpl(lock, worker, database, chunk);
             }
         }
+    }
 
-    } else {
+    LOGS(_log, LOG_LVL_DEBUG, context << "old replicas remov: 2");
```

Comment:
remov->remove ?

Reply:
This must stay 'remov' to have the same number of symbols as 'inser', and 'group' and 'compa' which are used for debugging purposes only. This makes counters reported by the debug messages properly aligned.